src/identity.rs

```diff
@@ -0,0 +1,844 @@
+        let mut filtered = BTreeMap::new();
+        for (pod_id, node_id) in &self.mapping {
+            if role == pod_id.role() && pod_id.group() == group {
+                filtered.insert(pod_id.clone(), node_id.clone());
+            }
+        }
+        filtered
+    }
+
+    pub fn get(&self, pod_id: &PodIdentity) -> Option<&NodeIdentity> {
+        self.mapping.get(pod_id)
+    }
+
+    pub fn insert(&mut self, pod_id: PodIdentity, node_id: NodeIdentity) -> Option<NodeIdentity> {
+        self.mapping.insert(pod_id, node_id)
+    }
+
+    pub fn filter(&self, id: &PodIdentity) -> Vec<NodeIdentity> {
+        self.mapping
+            .iter()
+            .filter_map(|(pod_id, node_id)| {
+                if pod_id.app() == id.app()
+                    && pod_id.instance() == id.instance()
+                    && pod_id.role() == id.role()
+                    && pod_id.group() == id.group()
+                {
+                    Some(node_id.clone())
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    pub fn merge(&self, other: &Self) -> Self {
+        let mut temp = self.mapping.clone();
+        temp.extend(other.clone().mapping);
+        PodToNodeMapping { mapping: temp }
+    }
+
+    /// Return true if the `node` is already mapped by a pod from `role` and `group`.
+    pub fn mapped_by(&self, node: &NodeIdentity, role: &str, group: &str) -> bool {
+        for (pod_id, mapped_node) in self.mapping.iter() {
+            if node == mapped_node && pod_id.role() == role && pod_id.group() == group {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Given `pods` return all that are not mapped.
+    pub fn missing(&self, pods: &[PodIdentity]) -> Vec<PodIdentity> {
+        let mut result = vec![];
+        for p in pods {
+            if !self.mapping.contains_key(p) {
+                result.push(p.clone())
+            }
+        }
+        result
+    }
+
+    #[cfg(test)]
+    pub fn new(map: Vec<(PodIdentity, NodeIdentity)>) -> Self {
+        let mut result = BTreeMap::new();
+        for (p, n) in map {
+            result.insert(p.clone(), n.clone());
+        }
+        PodToNodeMapping { mapping: result }
+    }
+}
+
+/// A pod identity generator that can be implemented by the operators.
+///
+/// Implementation of this trait are responsible for:
+/// - generating all pod identities expected by the service by implementing `PodIdentityFactory::as_slice`
+/// - map pods to their identity values by implementing `PodIdentityFactory::try_mapping`
+pub trait PodIdentityFactory {
+    /// A slice with all pod identities that should exist for a given service.
+    fn as_slice(&self) -> &[PodIdentity];
+    /// Returns a PodToNodeMapping for the given pods or an error if any pod could not be mapped.
+    fn try_from(&self, pods: &[Pod]) -> Result<Vec<PodIdentity>, Error>;
+}
```

Comment:
These both are existing traits. I would either make a "super" trait consisting of both the traits or just implement them one by one?

Reply:
I tried to factor out this trait into `AsRef<&[PodIdentity]>` and `TryFrom<&[Pod]> `but it turns out, the second one is not possible with the current design because our implementation of `try_from()` is not an associated function. It needs `&self` for the id label field and so on. We could rename it to avoid confusion.

Reply:
Renamed `try_from` to `try_map` to avoid confusion and made this trait a sub-trait of `AsRef<[PodIdentity]>`.